src/tooltip.js

```diff
@@ -106,4 +107,221 @@ class Tooltip {
 
 }
 
+class PopupManager {
+    constructor () {
+        this.popups = [];
+    }
+    
+    addPopup(popup) {
+        this.popups.push(popup);
+        this.updatePopups();
+    }
+
+    removePopup(popup) {
+        const index = this.popups.indexOf(popup);
+        if (index !== -1) {
+            this.popups.splice(index, 1);
+            this.updatePopups();
+        }
+    }
+
+    updatePopups() {
+        this.popups.sort((a, b) => b.priority - a.priority);
+        let visiblepopups = [];
+
+        for (let popup of this.popups) {
+            let shouldDisplay = true;
+            for (let visiblePopup of visiblepopups) {
+                if (this.doPopupsOverlap(visiblePopup, popup)) {
+                    shouldDisplay = false;
+                    break;
+                }
+            }
+            
+            if (shouldDisplay) {
+                visiblepopups.push(popup);
+            } else {
+                popup.hide();
+            }
+        }
+    }
+
+    doPopupsOverlap (popupA, popupB) {
+        const rectA = popupA.getRect();
+        const rectB = popupB.getRect();
+
+        return (rectA.left < rectB.right && rectA.right > rectB.left && rectA.top < rectB.bottom && rectA.bottom
+            > rectB.top);
+    }
+}
+
+var popupManager = new PopupManager();
+exports.popupManager = popupManager;
+
 exports.Tooltip = Tooltip;
+
+
+class HoverTooltip extends Tooltip {
+    constructor() {
+        super(document.body);
+        
+        this.timeout = undefined;
+        this.lastT = 0;
+        this.idleTime = 350;
+        this.lastEvent = undefined;
+        
+        this.onMouseOut = this.onMouseOut.bind(this);
+        this.onMouseMove = this.onMouseMove.bind(this);
+        this.waitForHover = this.waitForHover.bind(this);
+        this.hide = this.hide.bind(this);
+        
+        var el = this.getElement();
+        el.style.whiteSpace = "pre-wrap";
+        el.style.pointerEvents = "auto";
+        el.addEventListener("mouseout", this.onMouseOut);
+        el.classList.add("ace_doc-tooltip");
+        el.tabIndex = -1;
+        
+        el.addEventListener("blur", function() {
+            if (document.activeElement != el) this.hide();
+        }.bind(this));
+    }
+    
+    addToEditor(editor, callback, cancel) {
+        editor.on("mousemove", this.onMouseMove);
+    }
+    
+    onMouseMove(e, editor) {
```

Comment:
Seems like there are two issues with this method:
- it is not triggered when the mouse moves to the gutter (pointed out by @InspiredGuy)
- when moving the mouse beyond the last row and/or last column of the document that has content, `this.lastEvent.getDocumentPosition()` will keep returning the same values. Examples:
   - for a document containing a single token `hello`, the hover will never disappear unless you do a mouse click, whereas the expected behaviour is that hovering away in any direction from it would hide the tooltip
   - for a document containing multiple lines and then a line at the end containing `hello`, if you hover it, the tooltip does not disappear if you move the cursor to either:
      - the right, maintaining the cursor in the last row
      - downwards